data/categories/filter_categories.py

```diff
@@ -1,161 +1,107 @@
+    return [node for node in taxonomy.iter_nodes() if node_parent in node.get_parents_hierarchy()]
 
 def filter_categories(categories, parent_categories):
-    # get child nodes of parent_categories
-    node_child_list = list()
-    for parent_node in parent_categories:
-        parent_node_children = get_taxonomy_node_children_full_list(categories, parent_node)
-        node_child_list.extend(parent_node_children)
-
-    # add extra nodes
+    # Get all descendants for every parent node in parent_categories
+    all_descendants = [descendant for parent_node in parent_categories for descendant in get_all_descendants(categories, parent_node)]
+    # Add extra children to the list of descendants
     print("Add extra nodes:", EXTRA_CHILDREN)
-    for id in EXTRA_CHILDREN:
-        node = get_taxonomy_node_by_id(categories, id)
-        if node:
-            node_child_list.append(node)
-
-    # additional filtering
+    all_descendants.extend(get_taxonomy_node_list_by_id_list(categories, EXTRA_CHILDREN))
     print("Additional filtering on:", ADDITIONAL_FILTERING)
-    node_child_list_filtered = list()
-    for node in node_child_list:
-        if not node.get_localized_name("en").startswith(tuple(ADDITIONAL_FILTERING)):
-            node_child_list_filtered.append(node)
-
-    # remove duplicates
-    node_child_list_filtered_unique = list(set(node_child_list_filtered))
-
-    # keep only ids starting with "en:"
-    node_child_list_filtered_unique = [node for node in node_child_list_filtered_unique if node.id.startswith("en:")]
-
-    return node_child_list_filtered_unique
+    # additional filtering using regex word boundary that only keeps ids starting with "en:"
+    filtered_descendants = {node for node in all_descendants if not any(re.search(r'\b{}\b'.format(filter_word), node.get_localized_name("en"), flags=re.IGNORECASE) for filter_word in ADDITIONAL_FILTERING) and node.id.startswith("en:")}
+    return filtered_descendants
 
-
-def write_categories_to_files(categories):
+def write_categories_to_files(categories, delete_parents=False):
     languages = get_languages()
     for language in languages:
         language_code = language['code']
-        language_categories = list()
         # for each category, get translation (or default to en)
-        for category in categories:
-            language_category_name = category['name'][language_code] if (language_code in category['name']) else category['name']['en']
-            language_categories.append({"id": category['id'], "name": language_category_name})
+        language_categories = [{"id": category['id'], "name": category['name'].get(language_code, category['name']['en']), "parents": category.get('parents')} for category in categories]
+        # handle parents key
+        for i, category in enumerate(language_categories):
+            if not category["parents"] or delete_parents:
+                del language_categories[i]["parents"]
```

Comment:
why do you delete parent categories ? do you know which ones it actually deleted ?

Reply:
the risk is that sometimes the contributor doesn't have the info on which specific kind of apple or banana it is. but maybe the parent is much higher ?

Reply:
Because I'm adding the `parents` key in the list of dicts on the line before, using  `category.get('parents')`. The expected value then is either the list or `None`. It deleted the key if it has `None`, but also for any other `False` values such as an empty list if that ever happens. The other condition is a reimplementation of the `delete_parent` argument in the `taxonomy_node_list_to_dict_list` function in the old code. It was set to `delete_parent=True` but I wasn't really sure what the intention behind it was, so I re-implemented the behavior as it was originally. Was probably just added to reduce the file size of the generated JSON files, as the `parents` key is probably not used for anything anyway.
Here's a list of every unique parent category that loop deletes (156 in total):
```
["en:acidic-apples", "en:almonds", "en:apples", "en:apricots", "en:aromatic-plants", "en:artichokes", "en:asparagus", "en:avocados", "en:bananas", "en:basil", "en:bay-laurel", "en:berries", "en:bilberries", "en:black-grapes", "en:blueberries", "en:bouquet-garni", "en:brazil-nuts", "en:british-fruit", "en:broad-beans", "en:cabbages", "en:canned-tomatoes", "en:cantaloupe-melons", "en:carrots", "en:cashew-nuts", "en:celery", "en:cereal-flakes", "en:cherries", "en:chestnuts", "en:chives", "en:citrus", "en:clementines", "en:coconuts", "en:confectioneries", "en:coriander-leaves", "en:coriander-products", "en:culinary-plants", "en:dates", "en:dill", "en:dried-aromatic-plants", "en:dried-cherries", "en:dried-fruits", "en:dried-parsley", "en:dried-plant-based-foods", "en:elderberry-products", "en:fennel", "en:fenugreek-leaves", "en:fenugreek-products", "en:freeze-dried-plant-based-foods", "en:french-fruits", "en:fresh-chervil", "en:fresh-oranges", "en:fresh-plums", "en:fresh-pomelos", "en:fruits", "en:garlic", "en:garlic-and-their-products", "en:grapes", "en:grated-carrots", "en:green-beans", "en:green-cabbage", "en:green-grapes", "en:grilled-almonds", "en:ground-dried-aromatic-plants", "en:ground-dried-mint", "en:ground-dried-vegetables", "en:hazelnuts", "en:herbal-teas", "en:herbes-de-provence", "en:herbs", "en:kiwifruits", "en:kumquats", "en:leaf-vegetables", "en:leeks", "en:legumes", "en:lemon-verbena", "en:lemons", "en:lettuces", "en:limes", "en:lyophilized-aromatic-plants", "en:macadamia-nuts", "en:mandarin-oranges", "en:mangoes", "en:melons", "en:mint", "en:mixed-herbs", "en:mixed-vegetables", "en:mushrooms-and-their-products", "en:muskmelons", "en:nectarines", "en:nuts", "en:okras", "en:onions", "en:onions-and-their-products", "en:oranges", "en:oregano", "en:parsley", "en:peaches", "en:peanuts", "en:pears", "en:pecan-nuts", "en:peeled-asparagus", "en:pickled-vegetables", "en:pine-nuts", "en:pineapple", "en:pistachios", "en:pitted-cherries", "en:pomelos", "en:potatoes", "en:potatoes-from-france", "en:potatoes-from-italy", "en:potatoes-from-spain", "en:potted-aromatic-plants", "en:potted-mint", "en:potted-plants", "en:radishes", "en:red-cabbage", "en:roasted-hazelnuts", "en:roasted-nuts", "en:roasted-pistachios", "en:rolled-flakes", "en:rosemary", "en:russet-apples", "en:sage", "en:salsifis", "en:salted-almonds", "en:salted-nuts", "en:salty-snacks", "en:savory", "en:seedless-grapes", "en:seeds", "en:shelled-hazelnuts", "en:shelled-nuts", "en:shelled-seeds", "en:spanish-vegetables", "en:spearmint", "en:spices", "en:spinachs", "en:strawberries", "en:sweet-apples", "en:sweet-peppers", "en:tarragon", "en:textured-vegetable-protein", "en:thyme", "en:tomatoes", "en:tomatoes-and-their-products", "en:tropical-fruits", "en:turnip", "en:vegetable-rods", "en:vegetables", "en:vegetables-based-foods", "en:verbena-family", "en:walnuts", "en:white-asparagus", "en:whole-almonds", "en:zucchini", "fr:aubergines-preparees"]
```
Only 36 of these parents categories are not part of the final list:
```
["en:canned-tomatoes", "en:cereal-flakes", "en:confectioneries", "en:coriander-products", "en:culinary-plants", "en:dried-fruits", "en:dried-plant-based-foods", "en:elderberry-products", "en:fennel", "en:fenugreek-products", "en:freeze-dried-plant-based-foods", "en:fresh-chervil", "en:fresh-oranges", "en:fresh-plums", "en:fresh-pomelos", "en:fruits", "en:ground-dried-vegetables", "en:herbal-teas", "en:legumes", "en:mushrooms-and-their-products", "en:nuts", "en:onions-and-their-products", "en:pickled-vegetables", "en:potatoes", "en:potted-plants", "en:rolled-flakes", "en:salty-snacks", "en:seeds", "en:shelled-seeds", "en:spices", "en:textured-vegetable-protein", "en:tomatoes-and-their-products", "en:vegetables", "en:vegetables-based-foods", "en:verbena-family", "fr:aubergines-preparees"]
```